### ISSUE TITLE
Modal behind backdrop when parent is position:relative with z-index

### ISSUE THREAD

**Issue by yulunli** [twbs/bootstrap#16148]
When I use 3.4 to load a modal, the modal is covered by the backdrop. After removing all the customized css and leaving only the bootstrap css, the problem remains. Changing 3.4 to 3.2 instantly solve the problem. The modal is used within an ng-view partial.

**Comment by cvrebert** [comment]
We can't help unless you post an Angular-free JS Bin that demonstrates the problem.

**Comment by DiegoYungh** [comment]
New release 3.3.4 the modal backdrop is being left behind.
The call to modal('hide') is not doing it properly somehow.
Check that.

**Comment by cvrebert** [comment]
@DiegoYungh Again, we need a JS Bin example. The demos on getbootstrap.com seem to be working fine.

**Comment by nonameolsson** [comment]
@DiegoYungh You should post a code sample as the others ones already said..
I just want to mention that we've also had this problem. It was because of some bad overriding. It was because of some changes in the Bootstrap modal when upgrading from 3.3.2 -> 3.3.4.

Please post a code example and we'll figure it out.

**Comment by lustdante** [comment]
I can confirm this issue, although I don't have JS Bin example to show at the moment.

**Comment by lustdante** [comment]
https://gist.github.com/lustdante/86e2f1a1485a812c5a3a

Switch between 3.3.2 <-> 3.3.4 and you can replicate the issue.
It appears that 3.3.4 version render modal backdrop differently.

**Comment by twbs-lmvtfy** [comment]
Hi @lustdante!

You appear to have posted a live example (https://gist.github.com/lustdante/86e2f1a1485a812c5a3a), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W002](https://github.com/twbs/bootlint/wiki/W002): `<head>` is missing X-UA-Compatible `<meta>` tag that disables old IE compatibility modes
* [W003](https://github.com/twbs/bootlint/wiki/W003): `<head>` is missing viewport `<meta>` tag that enables responsiveness

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by cvrebert** [comment]
@lustdante In what OS and browser does the problem occur?

**Comment by lustdante** [comment]
OS X 10.10.2 + Chrome Version 41.0.2272.104 for now.

**Comment by lustdante** [comment]
Updated the gist. JSBin export screwed some settings.

https://gist.github.com/lustdante/86e2f1a1485a812c5a3a

**Comment by cvrebert** [comment]
JS Fiddle (since JS Bin has apparently run out of space): http://jsfiddle.net/1v84tto6/2/

**Comment by cvrebert** [comment]
> Changing 3.4 to 3.2

@yulunli You mean 3.3.4 to 3.3.2, right?

**Comment by cvrebert** [comment]
I can confirm @lustdante's issue.
Historical note: v3.3.0–v3.3.2 don't suffer from the problem, but v3.2.0 does.

**Comment by romant** [comment]
+1 on confirming the issue

… lesson learnt on relying on the latest - even if patch-only build.

**Comment by yulunli** [comment]
@cvrebert Yes. From 3.3.4 to 3.3.2. Sorry for the late response.

**Comment by yulunli** [comment]
I have reproduced the issue. You need to checkout out the repo and run a local node server to see it. https://github.com/yulunli/modal_bug

**Comment by leftclickben** [comment]
:+1: This is definitely an issue.  #15881 changes the DOM insertion for the backdrop element from `.prependTo(this.$element)` to `.appendTo(this.$body)`, this is the breaking change.

I'm not clear on whether a fiddle is still required for this?  If someone wants to advise I can put something up if needed.  As stated in original issue title, it only requires that the body has positioning and a z-index lower than the modal backdrop.

Will stick with 3.3.2 for now.

**Comment by benstr** [comment]
Yes I can confirm @leftclickben findings. It happened with this commit:

https://github.com/twbs/bootstrap/commit/f5beebe726aa8c1810015d8c62931f4559b49664#diff-452a8aa82569876ba89deba84331011cR194

**Comment by cvrebert** [comment]
Thanks folks; we don't need any further confirmations or fiddles.

**Comment by novtor** [comment]
+1. In my case modal closes but modal overlay stays shown. Firefox, bootstrap 3.3.4

**Comment by karlbateman** [comment]
This is driving me crazy, switching to version 3.3.2 still has no effect. Having the navbar-static-top omitted from the navbar makes no difference. Is there a quick fix for this?
![4267674595-capture](https://cloud.githubusercontent.com/assets/6918791/7322167/4d0cbed2-ea9e-11e4-8f24-6ef264a93227.png)

**Comment by sebcl** [comment]
same here, it's frustrating

**Comment by thomasphorton** [comment]
I'm looking into this now, just wanted to drop off a few notes in case anyone else is working on this.

@leftclickben identified the breaking PR correctly as #15881, but it's not the prepend/append change- I'm pretty sure it's the addition of `z-index: @zindex-modal-background;` in `less/modals.less`.

The z-index of the modal background is now 1040- so it makes sense that giving the wrapper a z-index of 1000 is going to create a stacking issue. I've got to brush up on my z-index rules though, because what I think will work isn't doing quite what I want.

**Comment by thomasphorton** [comment]
Quick response- it just took me typing that up to remember what I was looking for- the phrase is 'z-index stacking context'.

https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Understanding_z_index/The_stacking_context

By giving the wrapper element `position: relative`, we're creating a separate stacking context at `z-index: 1000`.

To solve this, I believe that `.modal-backdrop` should be generated as a sibling of `.modal` so that they can share z-index context, and resolve correctly. I haven't looked into the JS yet to see how simple this is.

**Comment by thomasphorton** [comment]
Fix is simple in the JS, but it looks like there's been a lot of turnover recently in how `.modal-background` is appended/prepended to various elements. I'm trying to gather some more information to make sure my solution doesn't cause any regressions.

**Comment by leftclickben** [comment]
@thomasphorton as you stated, the problem is the position within the DOM of the backdrop.

In 3.3.2, the `.modal-backdrop` is a sibling to the `.modal`, and this works correctly.

In 3.3.4, the `.modal-backdrop` is a child of the `body`, and this does not work correctly because the `z-index` context is incorrect.

It's good that you are looking into possible regressions here, it's a shame the person who made the change in 3.3.4 didn't do the same.

**Comment by thomasphorton** [comment]
@leftclickben yeah, it happens though. I'm not sure how that could have been caught outside of visual diff testing. Something to think about for the future.

**Comment by hnrch02** [comment]
@leftclickben I just want to mention this here again, the z-index problem was not intentionally fixed by #14724 (which shipped in 3.3.0) since it hasn't been reported before and therefore I had no knowledge of the erroneous behavior while I reverted #14724. This has been a problem since the inception of the modal plugin as the backdrop has always been appended to the body and not the modal element itself (as a little bit of digging through the history could've told you).

**Comment by hnrch02** [comment]
@leftclickben Also,

> In 3.3.2, the `.modal-backdrop` is a sibling to the `.modal`, and this works correctly.

this is incorrect. In 3.3.2 the `.modal-backdrop` is a sibling to the `.modal-dialog` and a direct descendant of the `.modal` element.

**Comment by hnrch02** [comment]
> I'm not sure how that could have been caught outside of visual diff testing.

Even if we had visual diffs it couldn't have been caught as the problem was reported long after #15881 happened.

**Comment by leftclickben** [comment]
@hnrch02 Sorry for my previous post, I was grumpy and I shouldn't have posted that.  I respect the work everyone does on twbs, it's an insanely difficult collection of use cases and support scenarios.  And yes, you are right, I meant "child" of `.modal`, not "sibling".

In terms of history and testing, I think the issue is that the behaviour in 3.3.4 is (in some cases) incompatibly different from the behaviour in 3.3.0, 3.3.1 and 3.3.2.  

In my mind, this type of change -- specifically moving an element from inside the widget's container to outside the widget's container, and possibly extended to any changes that affect the DOM element structure within the container -- should not be allowed in a patch version, because it is a point where user code interacts with library code (via `z-index` and other CSS properties placed on the widget's ancestors).

Visual diffs are not required -- the possibility of breaking BC should be enough to prevent a change like this, outside of a minor or major version change.

Anyway that is just my $.02 and I realise it doesn't help here, nor am I otherwise a contributor to twbs, but maybe something to consider for coding and/or version control guidelines.

**Comment by mdo** [comment]
Punting per #16432.

**Comment by RickStrahl** [comment]
What is the status of this? I see the issue closed but I still see the problem in 3.3.5 - same issue as others I have an absolutely positioned container overlays everything even when then modal's z-index is higher.

**Comment by cvrebert** [comment]
@RickStrahl Per the linked pull request, this is WontFix for Bootstrap v3.

**Comment by thomasphorton** [comment]
@RickStrahl it's not ideal, but the fix is pretty easy if you follow my PR. I don't know if v3 is at EOL or not, but maybe you can fix it the one time and wait for v4 for your next upgrade?
